### PR TITLE
fix checkLclVarSemantics

### PR DESCRIPTION
**Pull request by sandreenko**
Fix checkLclVarSemantics logic to check consumed nodes writes and reads correctly.

It is the first part of #13910([RyuJit, JitStress] Set contained ignores lclVars stores).

### CODE REVIEW

**Comment by sandreenko** [comment]
The first PR just refactor the existing checker.

**Comment by sandreenko** [comment]
The second should fail with a bad execution.

**Comment by sandreenko** [comment]
@dotnet test Windows_NT x64 Debug Build and Test

**Comment by sandreenko** [comment]
The second PR shows the silent bad code generation.

**Comment by sandreenko** [comment]
The third PR fixes the checker and ends up with assert on the test.

**Comment by sandreenko** [comment]
PTAL @CarolEidt, @dotnet/jit-contrib.

**Comment by sandreenko** [comment]
Please review each commit separately.

**Comment by sandreenko** [comment]
The fourth commit will add the test to the exclusion list, where it will stay until I fix the second part of the original problem (I have not found a fix that works fast enough and is not over conservative yet).

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Runtime.InteropServices;
+
+// Represents a problem with contained nodes chaines, that contain lclVar reads, that were moved through lclVar stores.
```
Comment (BruceForstall) [inline tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: typo: chaines => chains

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    private static int test_2_2(int num)
+    {
+        int result;
+        if (AA.a_init[num].q != AA.a_zero[num].q) 
+        // Access field with contained IND instruction.
+        // EQ marks its operands as contained too. 
+        // AA.a_init[num].q and AA.a_zero[num].q are allocated to the same lclVar.
+        // So we calculate AA.a_init[num].q and store as tmp0, use this temp to do nullCheck.
+        // Then store AA.a_zero[num].q as tmo0, destroy the old value and try to do EQ thinking that
```
Comment (BruceForstall) [inline tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: tmo0 => tmp0

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        int result;
+        if (AA.a_init[num].q != AA.a_zero[num].q) 
+        // Access field with contained IND instruction.
+        // EQ marks its operands as contained too. 
+        // AA.a_init[num].q and AA.a_zero[num].q are allocated to the same lclVar.
+        // So we calculate AA.a_init[num].q and store as tmp0, use this temp to do nullCheck.
+        // Then store AA.a_zero[num].q as tmo0, destroy the old value and try to do EQ thinking that 
+        // tmp0 is AA.a_init[num].q.
+        // It needs stress (complus_JitStressModeNames=STRESS_NULL_OBJECT_CHECK, STRESS_MAKE_CSE)
+        // to force the compiler to do implicit null checks and store values as local variables.
```
Comment (BruceForstall) [inline tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: would it be worthwhile to include an example of the bad LIR?

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.csproj [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
+  </ItemGroup>
+  <PropertyGroup>
+    <CLRTestBatchPreCommands><![CDATA[
+$(CLRTestBatchPreCommands)
+set complus_JitStressModeNames=STRESS_NULL_OBJECT_CHECK, STRESS_MAKE_CSE
+]]></CLRTestBatchPreCommands>
+    <BashCLRTestPreCommands><![CDATA[
+$(BashCLRTestPreCommands)
+export complus_JitStressModeNames=STRESS_NULL_OBJECT_CHECK, STRESS_MAKE_CSE
```
Comment (BruceForstall) [inline tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.csproj]: You need `complus` to be properly capitalized: `COMPlus`, since otherwise, on Linux, you won't be setting the correct (case-sensitive) variable. I would write it this way everywhere, including in comments, just to be clear.

Comment (sandreenko) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.csproj]: Fixed

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 2 above)
Comment (sandreenko) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: Fixed

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 1 above)
Comment (sandreenko) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: Fixed

File: src/jit/lir.cpp [hunk 5]
```diff
@@ -1407,6 +1407,92 @@ LIR::ReadOnlyRange LIR::Range::GetRangeOfOperandTrees(GenTree* root, bool* isClo
...
+            // If this node is a lclVar write, it must be to a lclVar that does not have an outstanding read.
+            assert(!nodeInfo.IsLclVarWrite() || !unconsumedLclVarReads.Contains(nodeInfo.LclNum()));
+        }
+
+        return true;
+    }
+
+private:
+    //------------------------------------------------------------------------
+    // ConsumeNodeOperands: mark the node's operands as consumed.
```
Comment (CarolEidt) [inline src/jit/lir.cpp]: I don't think this should talk about consuming node operands. The term "consume" has specific meaning in the backend, and it appears that this is using that term to mean something different.

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 1 above)
Comment (CarolEidt) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: The comment above is very difficult to parse, and it is unclear what it means to move a lclVar read through a lclVar store. Please show an IR excerpt or something so it's clearer what the problem is.

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 2 above)
Comment (CarolEidt) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: It is not clear why `AA.a_init[num].q` and `AA.a_zero[num].q` would be allocated to the same lclVar; I presume that the JIT is doing some reuse of struct temps?

File: src/jit/lir.cpp (hunk 5 above)
Comment (sandreenko) [reply src/jit/lir.cpp]: Il example was added.

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 1 above)
Comment (sandreenko) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: Il example was added in the last commit.

File: src/jit/lir.cpp (hunk 5 above)
Comment (sandreenko) [reply src/jit/lir.cpp]: I used the same naming as it was before in `unconsumedLclVarReads`. Do we want rename `unconsumedLclVarReads` as well?

Comment (CarolEidt) [reply src/jit/lir.cpp]: > Do we want rename unconsumedLclVarReads as well?

Yes, I think so. That is not as bad, because it is not claiming to consume them, which has implications for liveness updates, etc. But I think it's bad practice to refer to consuming at all in this context.

Comment (sandreenko) [reply src/jit/lir.cpp]: I think what we do is actually  liveness analysis here.
I can't find a better name for it, but if you have one I will rename `unconsumedLclVarReads` and `ConsumeNodeOperands` to that.

File: src/jit/decomposelongs.cpp [hunk 6]
```diff
@@ -113,8 +113,6 @@ void DecomposeLongs::DecomposeRangeHelper()
     {
         node = DecomposeNode(node);
     }
-
-    assert(Range().CheckLIR(m_compiler, true));
```
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: Why was this removed?

File: src/jit/lir.cpp (hunk 5 above)
Comment (pgavlin) [reply src/jit/lir.cpp]: We aren't quite doing liveness here, though we are doing something somewhat related. What we are trying to do in ensure that given a node `t0` that represents a use of a lclVar `V0` and its using node `tX`, there are no defs of `V0` between `t0` and `tX`. This is a requirement of our model wherein lclVar uses occur at the user rather than at the lclVar node.

I think that we could use `unusedLclVarReads` and `UseNodeOperands` in place of `unconsumedLclVarReads` and `ConsumeNodeOperands`.

File: src/jit/decomposelongs.cpp (hunk 6 above)
Comment (sandreenko) [reply src/jit/decomposelongs.cpp]: We have two phases: DecomposeBlock and LowerBlock. This check was called between them.
The contained flag doesn't have to be correctly set before we call `LowerBlock()`. 
We have an assert that `GT_LONG` is contained or doesn't have a user in this `CheckLIR`.
So this check failed on x86.

File: src/jit/lir.cpp (hunk 5 above)
Comment (sandreenko) [reply src/jit/lir.cpp]: Fixed.

File: src/jit/decomposelongs.cpp (hunk 6 above)
Comment (pgavlin) [reply src/jit/decomposelongs.cpp]: Interesting. It would be nice if we could leave this in, as it has caught problems before. Where specifically is the check you mention?

Comment (sandreenko) [reply src/jit/decomposelongs.cpp]: It is in `isContained`:
```
#if !defined(_TARGET_64BIT_)
    if (OperGet() == GT_LONG)
    {
        // GT_LONG nodes are normally contained. The only exception is when the result
        // of a TYP_LONG operation is not used and this can only happen if the GT_LONG
        // has no parent.
        assert(isMarkedContained || (gtGetParent(nullptr) == nullptr));
    }
#endif
```
This check is useful so it seems unlikely to skip it when we call CheckLIR from the intermediate state.

**Comment by sandreenko** [comment]
I merged PR#14044: mark argplace node as no_lir. Need some time to fix new failures here.

**Comment by sandreenko** [comment]
PR was updated. Please take a look.

File: tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs (hunk 2 above)
Comment (sandreenko) [reply tests/src/JIT/Regression/JitBlue/CoreCLR_13910/CoreCLR_13910.cs]: > It is not clear why AA.a_init[num].q and AA.a_zero[num].q would be allocated to the same lclVar; I presume that the JIT is doing some reuse of struct temps?

Yes

File: src/jit/decomposelongs.cpp (hunk 6 above)
Comment (pgavlin) [reply src/jit/decomposelongs.cpp]: I think that we should move this check out of `isContained` and into whatever phase expects this behavior (likely codegen). This should be simple: there are already cases for `GT_LONG` in codegen{x,arm}arch.cpp, so the requisite assert (namely, that an uncontained `GT_LONG` encountered during codegen has a user) could easily be added there. This would clarify the requirement a bit by making it obvious which phase expects this behavior and would allow us to keep checking LIR correctness between decomp and lowering.

Comment (sandreenko) [reply src/jit/decomposelongs.cpp]: It sounds reasonable. 
`genConsumeRegs` is the place, where we expect this flag to be set.
I updated PR with the fix.

**Comment by CarolEidt** [comment]
> genConsumeRegs is the place, where we expect this flag to be set.
I updated PR with the fix.

That is not correct. This should be checked in the register allocator, specifically in `LinearScan::TreeNodeInfoInit()`, which is where the containment settings are first used (and expected to be correct). This code: https://github.com/dotnet/coreclr/blob/master/src/jit/lsraxarch.cpp#L204 and the corresponding code in lsraarm.cpp should be changed to an assert, since that code should not be reached for a `GT_LONG` that is not unused (there is a containment check at the beginning of the method that does an early return).

**Comment by sandreenko** [comment]
> That is not correct. This should be checked in the register allocator, specifically in LinearScan::TreeNodeInfoInit(), which is where the containment settings are first used (and expected to be correct). This code: https://github.com/dotnet/coreclr/blob/master/src/jit/lsraxarch.cpp#L204 and the corresponding code in lsraarm.cpp should be changed to an assert, since that code should not be reached for a GT_LONG that is not unused (there is a containment check at the beginning of the method that does an early return).

Thank you, fixed.

**Comment by sandreenko** [comment]
Thanks all, the second part is coming)
